tt-train/tests/model/nano_gpt_test.cpp

```diff
@@ -0,0 +1,235 @@
+            end_timer = std::chrono::high_resolution_clock::now();
+            duration = std::chrono::duration_cast<std::chrono::microseconds>(end_timer - start_timer).count();
+            fmt::print("dataloader step time {} ms\n", (double)duration / 1000.);
+            return std::make_tuple(data_tensor, targets_tensor, cached_data.masks_tensor, cached_data.positions_tensor);
+        };
+    auto train_dataloader = DataLoader(dataset, /* batch_size */ config.batch_size, /* shuffle */ true, collate_fn);
+
+    fmt::print("Overriding vocab size to be divisible by 32\n");
+    config.transformer_config.vocab_size = (tokenizer.get_vocab_size() + 31) / 32 * 32;
+    auto model = ttml::models::gpt2::create(config.transformer_config);
+
+    auto adamw_params = ttml::optimizers::AdamWConfig();
+    adamw_params.lr = config.learning_rate;
+    adamw_params.weight_decay = config.weight_decay;
+
+    auto create_optimizer = [&]() -> std::shared_ptr<ttml::optimizers::OptimizerBase> {
+        if (use_moreh_adamw) {
+            return std::make_shared<ttml::optimizers::MorehAdamW>(model->parameters(), adamw_params);
+        } else {
+            return std::make_shared<ttml::optimizers::AdamW>(model->parameters(), adamw_params);
+        }
+    };
+
+    auto optimizer = create_optimizer();
+
+    std::vector<double> steps_time;
+    std::vector<float> losses;
+
+    for (auto [features, target, masks, positions] : train_dataloader) {
+        auto start_timer = std::chrono::high_resolution_clock::now();
+        optimizer->zero_grad();
+        auto output = (*model)(features, positions, masks);
+        auto loss = ttml::ops::nll_loss(output, target);
+        auto loss_float = ttml::core::to_vector(loss->get_value())[0];
+        loss->backward();
+        optimizer->step();
+        ttml::autograd::ctx().reset_graph();
+        auto global_step = optimizer->get_steps();
+        losses.emplace_back(loss_float);
+        if (global_step >= config.max_steps) {
+            break;
+        }
+        auto end_timer = std::chrono::high_resolution_clock::now();
+        auto duration = std::chrono::duration_cast<std::chrono::microseconds>(end_timer - start_timer).count();
+        steps_time.emplace_back((double)duration / 1000.0);
+    }
+
+    // verify program cache
+    auto program_cache_entries = device->num_program_cache_entries();
+    if (!use_moreh_adamw) {
+        EXPECT_EQ(program_cache_entries, 124);
+    } else {
+        EXPECT_EQ(program_cache_entries, 103);
+    }
+
+    // verify time per step
+    size_t num_steps_below = 0;
+    double expected_time_ms = 330.0;
+    for (auto &time : steps_time) {
+        num_steps_below += (time < expected_time_ms);
+    }
+    if (num_steps_below / static_cast<double>(steps_time.size()) < 0.9) {
+        EXPECT_TRUE(false);
+    }
+
+    // verify loss
+    if (!use_moreh_adamw) {
```

Comment:
overall I'd add a huge comment that it is nanogpt training regression tests. If they are broken it means:
1) op accuracy changes
2) new ops added silently
3) just bug introduced

Reply:
Sounds good! Will do

Reply:
Added!